core/admin/mailu/utils.py

```diff
@@ -48,3 +71,359 @@ def init_app(self, app):
+            return None
+        return b''.join([self._uid, self._sid, self._created])
+
+    def destroy(self):
+        """ destroy session for security reasons. """
+
+        self.delete()
+
+        self._uid = None
+        self._sid = None
+        self._created = None
+
+        self.clear()
+
+        self.modified = True
+        self.new = False
+
+    def regenerate(self):
+        """ generate new id for session to avoid `session fixation`. """
+
+        self.delete()
+
+        self._sid = None
+        self._created = self.app.session_config.gen_created()
+
+        self.modified = True
+
+    def delete(self):
+        """ Delete stored session. """
+        if self.saved:
+            self.app.session_store.delete(self._key)
+            self._key = None
+
+    def save(self):
+        """ Save session to store. """
+
+        # set uid from dict data
+        if self._uid is None:
+            self._uid = self.app.session_config.gen_uid(self.get('user_id', ''))
+
+        # create new session id for new or regenerated sessions
+        if self._sid is None:
+            self._sid = self.app.session_config.gen_sid()
+
+        # get new session key
+        key = self.sid
+
+        # delete old session if key has changed
+        if key != self._key:
+            self.delete()
+
+        # save session
+        self.app.session_store.put(
+            key,
+            pickle.dumps(dict(self)),
+            self.app.permanent_session_lifetime.total_seconds()
+        )
+
+        self._key = key
+
+        self.new = False
+        self.modified = False
+
+class MailuSessionConfig:
+    """ Stores sessions crypto config """
+
+    def __init__(self, app=None):
+
+        if app is None:
+            app = flask.current_app
+
+        bits = app.config.get('SESSION_KEY_BITS', 64)
+
+        if bits < 64:
+            raise ValueError('Session id entropy must not be less than 64 bits!')
+
+        hash_bytes = bits//8 + (bits%8>0)
+        time_bytes = 4 # 32 bit timestamp for now
+        shaker = hashlib.shake_256 if bits>128 else hashlib.shake_128
+
+        self._shaker   = shaker(want_bytes(app.config.get('SECRET_KEY', '')))
+        self._hash_len = hash_bytes
+        self._hash_b64 = len(self._encode(bytes(hash_bytes)))
+        self._key_min  = 2*self._hash_b64
+        self._key_max  = self._key_min + len(self._encode(bytes(time_bytes)))
+
+    def gen_sid(self):
+        """ Generate random session id. """
+        return self._encode(secrets.token_bytes(self._hash_len))
+
+    def gen_uid(self, uid):
+        """ Generate hashed user id part of session key. """
+        shaker = self._shaker.copy()
+        shaker.update(want_bytes(uid))
+        return self._encode(shaker.digest(self._hash_len))
+
+    def gen_created(self, now=None):
+        """ Generate base64 representation of creation time. """
+        return self._encode(int(now or time.time()).to_bytes(8, byteorder='big').lstrip(b'\0'))
```

Comment:
Can we make the time UTC everywhere? It may save us headaches down the line if people cluster it across different timezones

Reply:
turns out that time.time() is UTC already.